### ISSUE TITLE
swagger.json is using int64 for things which should be uint64

### ISSUE THREAD

**Issue by randomshinichi** [aeternity/aeternity#2120]
## Expected Behavior
Key parameters such as KeyBlockHeight, MicroBlockHeight, amount target, nonce etc. should be expressed in big integers, or at least `uint64`, not `int64`.

~I've edited swagger.json to use uint64 and bigint for Amount [swagger.json](https://github.com/aeternity/aepp-sdk-go/blob/removeconfig2/api/swagger.json) and the [custom validation needed for bigint](https://github.com/aeternity/aepp-sdk-go/commit/553c6b43efae6817d963fde81e0a961784a23fac#diff-db8f6e777af95876086f2afbb40aa3a9R58).~
Please use the files in my reply to this comment.

Ideally uint64 and ~bigints~ an "Amount" model should already be in the swagger.json and I can just add the custom validation function by hand.
...

## Actual Behavior
swagger.json always has int64s.
...

## Steps to Reproduce the Problem

  1. `curl  http://localhost:3013/api -o api/swagger.json`
  
## Logs, error output, etc.
This causes problems with golang, which is stricter than Javascript or Python about such things.
This means that I have to manually modify swagger generated code too, which is not good.

## Specifications

  - Virtualization: [*none]
  - Hardware specs: 
    * RAM: 16GB
    * CPU: Intel Ivy Bridge
  - OS: [Manjaro]
  - Node Version: 1.4.0

**Comment by randomshinichi** [comment]
https://github.com/aeternity/aepp-sdk-go/blob/develop/api/swagger.json

Here I have converted almost every "integer" into uint64 except for time, where apparently [Golang thinks](https://golang.org/pkg/time/#Unix) is reasonable to express as a int64.

https://github.com/aeternity/aepp-sdk-go/blob/develop/api/edit.py
This is a Python script that describes the changes (not all of them) that I always make to swagger.json.

**Comment by marc0olo** [comment]
we faced the same problem when developing the java sdk:
- https://forum.aeternity.com/t/swagger-yml-and-code-generation/3155

**Comment by randomshinichi** [comment]
What's the status on this issue? It is slowing down integration of new node versions...

**Comment by hanssv** [comment]
Sadly the swagger format isn't as good as one could wish for... There is no way to indicate that a number is a 64-bit unsigned integer, etc. (And no, an `integer` with format `int64` and minimum `0` is not the same thing...) 

In the Aeternity API we have no signed integers, and we have 32-bit unsigned integers, 64-bit unsigned integers and unbound integers. Currently some of the unbound integers are marked with `format: int64` as well as most of the 64-bit integers which is just plain wrong.

My suggestion is to introduce three types:
```yaml
UInt32:
  type: integer
  minimum: 0
  maximum: 4294967295

UInt64:
  type: integer
  minimum: 0
  maximum: 18446744073709551615

UInt:
  type: integer
  minimum: 0
```

That is as good as we can do it (provided that we then use the right type for the right field of course!) - I'll see when we can get this done...

**Comment by randomshinichi** [comment]
```
UInt32:
  type: integer
  minimum: 0
  maximum: 4294967295

UInt64:
  type: integer
  minimum: 0
  maximum: 18446744073709551615

UInt:
  type: integer
  minimum: 0
```
These mins and maxs will result in swagger generating Validate() functions which further need to be disabled, because for Python/JS, the integers are unbounded and these checks would just get in the way. For Go, I would also end up disabling these checks (that's extra editing work) because they would result in misleading behaviour that looks like a user error, but actually is a result of the number hitting the limit.

Since the OAS specification looks so limited, we will need to tailor the swagger.json to our swagger code generators (Go to go-swagger, Java to whatever...), I would suggest just changing the swagger file so that it is easy to edit. This means:
1. every type: integer should have a format: int64
2. Perhaps introduce custom types for `Fee, Balance, Amount, NameSalt, Gas, GasPrice`. These types can be as simple as
```
    "Amount": {
      "type": "integer",
      "minimum": 0,
      }
    },
```
So that later it will be easier for Go to replace this with
```
    "Amount": {
      "type": "integer",
      "minimum": 0,
      "x-go-type": {
        "import": {
          "package": "github.com/aeternity/aepp-sdk-go/utils"
        },
        "type": "BigInt"
      }
    },
```

**Comment by hanssv** [comment]
Why in that case can't you just replace
```
UInt64:
  type: integer
  minimum: 0
  maximum: 18446744073709551615
```
with
```
UInt64:
  type: integer
  minimum: 0
  x-go-type:
    import:
      package: github.com/aeternity/aepp-sdk-go/utils
    type: BigInt
```

**Comment by hanssv** [comment]
In general I'm mostly concerned with the correctness and consistency of the `swagger.json` if it suits the tools it is good, but that is secondary.

**Comment by hanssv** [comment]
This should be addressed in #2437

**Comment by randomshinichi** [comment]
It turns out the Uint16/32/64/Uint/TxBlockHeight changes to swagger.json were very useful. Thank you!

However, I still noticed some small inconsistencies. The following paths still use a normal integer with 'limits' but should probably reference the Uint64 model instead:
```
/key-blocks/height/height
/generations/height/height
/accounts/pubkey/height/height
/debug/token-supply/height/height
```

`HeightIn/SaltIn` appears to be unused. What are they for?
